### ISSUE TITLE
[VMR] Design and introduce method of specifying the dotnet host

### ISSUE THREAD

**Issue by mmitche** [dotnet/dotnet#1008]
### Describe the Problem

Arcade has a set of functionality that is tasked with picking the dotnet host that should be used when executing a build command. How it is determined is based on global.json, available visual studio versions, platforms, etc. Linux source build has decided what dotnet executable is the host that should be used based the fact that the SDK. installed is always applocal, then put on the path. I (think) this is then transparently picked up by the arcade tooling. For the Windows PoC's, we've run into complexities in Roslyn, which is typically looking for a desktop msbuild.

### Describe the Solution

We should introduce standardized behavior into repo scripting (ideally arcade) which allows a build invocation to be explicit about the dotnet host that should be used, rather than assuming the host is on the path, or picking a VS install and using desktop msbuild.

**T-Shirt Size: Medium/Large (may not be critical)**

**Comment by steveisok** [comment]
Bringing over @jaredpar comments from https://github.com/dotnet/source-build/issues/3672

Couple of items that I think need to be solved:

build.sh / .cmd level: 
1. method of specifying `dotnet` 
2. method of controlling the sdk version (or are we just stomping global.json)
3. method to specify it's the vmr build so full product assets should be produced. Many builds for example don't produce .nupkg by default today
4. is restoring implicit when (3) is passed?
5. do we need to support debug / release type flags?
6. are we using build.cmd/sh in the root or will we have say `eng/vmr-build.cmd/sh`?
7. method of specifying binary logs (we are **really** going to want that so we can track down build issues in this complex build)

msbuild props level: 
1. prop similar to `$(DotNetBuildFromSource)` to indicate we're in a vmr build so we can condition the build. Example is roslyn probably conditions out all of their VS integration layers for a full VMR build
2. prop specifying the ideal `<TargetFramework>` version for the vmr build. Similar in function to `$(NetCurrent)`. 
3. do we need to push down ideal RIDs?

**Comment by mmitche** [comment]
> * method of controlling the sdk version (or are we just stomping global.json)

This is an interesting topic. We stomp it today, but this leads to changes in the sources. Controlling the SDK version outside of changing global.json will likely require product changes. This is from a chat with @rainersigwald.

**Comment by jaredpar** [comment]
> This is an interesting topic. We stomp it today, but this leads to changes in the sources. Controlling the SDK version outside of changing global.json will likely require product changes. This is from a chat with @rainersigwald.

This request has come in many times from many customers. It's a constant source of pain that have a `dotnet` in hand, with a corresponding SDK, and cannot tell `dotnet` please just build with this SDK. There is no other ecosystem out there that has this problem. 

I think we shoud really consider putting a top level option in `dotnet` that lets us control the SDK version and effectively ignore whatever is in `global.json`. Somethng akin to `dotnet --sdk x.y`. As long as `x.y` is available in `--list-sdks` then it is used. 

@baronfel

**Comment by steveisok** [comment]
What has prevented us from doing this in the past? Is it just the will or something else?

**Comment by mmitche** [comment]
I'd argue that you really want this for all of global.json. The ability to override any individual elements.

**Comment by jaredpar** [comment]
> What has prevented us from doing this in the past? Is it just the will or something else? 

@baronfel would have the definitive answer on that. 

My view is that there are some aspects that have to be designed. For example even if you had a `--ignore-global-json` flag that still requires design. A given `dotnet` can have multiple SDKS available so which one do you use: latest' earliest, etc ... That is why I used the `--sdk` flag above cause it gets right to that problem. 

My simplistic (and possibly incorrect) view is that it would be feasible to have someting akin to: 

- `dotnet --sdk`: use the latest SDK available in `dotnet --list-sdks` to build and ignore any sdk restrictions in `global.json`
- `dotnet --sdk x.y`: use SDK x.y to build. Issue an error if x.y is not available in `--list-sdks`

**Comment by jaredpar** [comment]
> I'd argue that you really want this for all of global.json. The ability to override any individual elements.

At that point you probably just want a global option for `--the-real-global-json` cause. That though puts you right back in the same situtaion you were in before: in order to build with the SDK you literally have ni hand and have referenced by a full path you **still** need to modify the file system.

**Comment by mmitche** [comment]
IIRC, @rainersigwald noted that multiple tools may read/use the global.json today (NuGet, msbuild, host?, VS...?) which may present a challenge to providing an override mechanism

**Comment by baronfel** [comment]
It's coordination and design - everything you're describing is actually implemented by the native dotnet/runtime host, not the SDK, so there's also a question of who owns the behavior. And then there's tooling impact - if you sidestep global.json how do all the other participants in the ecosystem make sure they're running the same way?

**Comment by jaredpar** [comment]
> IIRC, @rainersigwald noted that multiple tools may read/use the global.json today (NuGet, msbuild, VS...?) which may present a challenge to providing an override mechanism

We own all those tools and can fix them. 

Consider that VS is not a concern here. This is about how you run `msbuild / dotnet build` on the command line. Has no VS impact.

**Comment by mmitche** [comment]
> > I'd argue that you really want this for all of global.json. The ability to override any individual elements.
> 
> At that point you probably just want a global option for `--the-real-global-json` cause. That though puts you right back in the same situation you were in before: in order to build with the SDK you literally have ni hand and have referenced by a full path you **still** need to modify the file system.

In order for things like source-build/VMR build to work though, you need to be able to override more than just the dotnet SDK. MSBuild SDKs are specified within the global.json. Those version numbers need to be updated to achieve consistency at build time. You need to be able to tell it: use this version I just built.

So you're left with two options:
-  `--will-the-real-global-json-please-stand-up` - We would have the option of creating one on disk (not in the source tree, but in the artifacts), and telling all the tools to look there for it.
- `<option set of some complexity to provide general overrding of global.json functionality>`

**Comment by jaredpar** [comment]
> You need to be able to tell it: use this version I just built.

You should be able to build without mutating the build source. It's a fairly core build principal. We do not meet this principal in .NET.  We keep finding reasons to not fix this. 

Instead we keep inventing systems like: 

1. Download .NET 
2. Write something to the file system so I can use the .NET I just downloaded. 

No other ecosystem has this problem.

**Comment by mmitche** [comment]
Agreed. I just worry that `global.json` is sort of an arbitrary complex format. Allowing its full breadth to be specifiable on the command line may be awkward. Not saying we shouldn't, just that it may be weird. I think this says more about the design of `global.json` than anything else.

I mean, check out wpf's:

```
{
  "tools": {
    "dotnet": "9.0.100-alpha.1.23511.2",
    "runtimes": {
      "dotnet": [
        "2.1.7",
        "$(VSRedistCommonNetCoreSharedFrameworkx6490PackageVersion)"
      ]
    },
    "vs": {
      "version": "17.2"
    }
  },
  "msbuild-sdks": {
    "Microsoft.DotNet.Arcade.Sdk": "9.0.0-beta.23527.1",
    "Microsoft.DotNet.Helix.Sdk": "9.0.0-beta.23527.1"
  },
  "sdk": {
    "version": "9.0.100-alpha.1.23511.2"
  },
  "native-tools": {
    "strawberry-perl": "5.38.0.1",
    "net-framework-48-ref-assemblies": "0.0.0.1"
  }
}
```

**Comment by akoeplinger** [comment]
is this still needed now that we have  https://github.com/dotnet/designs/blob/main/accepted/2025/local-sdk-global-json.md?

**Comment by mmitche** [comment]
No, I don't think so...we would use that feature. We did build a bit of infra up to hook into the msbuildEngine <engine> parameter in the arcade build scripts.
